openssl/aes.go

```diff
@@ -342,6 +352,25 @@ func (g *aesGCM) Seal(dst, nonce, plaintext, additionalData []byte) []byte {
 	if len(dst)+len(plaintext)+gcmTagSize < len(dst) {
 		panic("cipher: message too large for buffer")
 	}
+	if g.tls {
+		if len(additionalData) != gcmTlsAddSize {
+			panic("cipher: incorrect additional data length given to GCM TLS")
+		}
+		// BoringCrypto enforces strictly monotonically increasing explicit nonces
+		// and to fail after 2^64 - 1 keys as per FIPS 140-2 IG A.5,
+		// but OpenSSL does not perform this check, so it is implemented here.
+		const maxUint64 = 1<<64 - 1
+		counter := bigUint64(nonce[gcmTlsFixedNonceSize:])
+		if g.minNextNonce == maxUint64 {
```

Comment:
I think this `if` either needs to (also) check `counter == maxUint64`, or `g.minNextNonce = counter + 1` needs bounds checking to prevent overflow. Right now when you pass in `1<<64 - 1` as the counter part of the nonce, it ends up setting `g.minNextNonce` to 0 so you can pass in whatever you want next time.

Reply:
You are right. I've fixed it and added a test. In this PR I've clearly invested more time in documentation than in validation 😞 .